lib/Dialect/Arc/Transforms/LowerStateRewrite.cpp

```diff
@@ -0,0 +1,1206 @@
+  LogicalResult lower(seq::InitialOp op);
+  LogicalResult lower(llhd::FinalOp op);
+
+  scf::IfOp createIfClockOp(Value clock);
+
+  // Value Lowering. These functions are called from the `lower()` functions
+  // above. They handle values used by the `op`. This can generate reads from
+  // state and memory storage on-the-fly, or mark other ops as dependencies to
+  // be lowered first.
+  Value lowerValue(Value value, Phase phase);
+  Value lowerValue(InstanceOp op, OpResult result, Phase phase);
+  Value lowerValue(StateOp op, OpResult result, Phase phase);
+  Value lowerValue(sim::DPICallOp op, OpResult result, Phase phase);
+  Value lowerValue(MemoryReadPortOp op, OpResult result, Phase phase);
+  Value lowerValue(seq::InitialOp op, OpResult result, Phase phase);
+  Value lowerValue(seq::FromImmutableOp op, OpResult result, Phase phase);
+
+  void addPending(Value value, Phase phase);
+  void addPending(Operation *op, Phase phase);
+};
+
+/// All state associated with lowering a single module.
+struct ModuleLowering {
+  /// The module being lowered.
+  HWModuleOp moduleOp;
+  /// The builder for the main body of the model.
+  OpBuilder builder;
+  /// The builder for state allocation ops.
+  OpBuilder allocBuilder;
+  /// The builder for the initial phase.
+  OpBuilder initialBuilder;
+  /// The builder for the final phase.
+  OpBuilder finalBuilder;
+
+  /// The storage value that can be used for `arc.alloc_state` and friends.
+  Value storageArg;
+
+  /// A worklist of pending op lowerings.
+  SmallVector<OpLowering> opsWorklist;
+  /// The set of ops currently in the worklist. Used to detect cycles.
+  SmallDenseSet<std::pair<Operation *, Phase>> opsSeen;
+  /// The ops that have already been lowered.
+  DenseSet<std::pair<Operation *, Phase>> loweredOps;
+  /// The values that have already been lowered.
+  DenseMap<std::pair<Value, Phase>, Value> loweredValues;
+
+  /// The allocated input ports.
+  SmallVector<Value> allocatedInputs;
+  /// The allocated states as a mapping from op results to `arc.alloc_state`
+  /// results.
+  DenseMap<Value, Value> allocatedStates;
+  /// The allocated storage for instance inputs and top module outputs.
+  DenseMap<OpOperand *, Value> allocatedOutputs;
+  /// The allocated storage for values computed during the initial phase.
+  DenseMap<Value, Value> allocatedInitials;
+  /// The allocated storage for taps.
+  DenseMap<Operation *, Value> allocatedTaps;
+
+  /// A mapping from unlowered clocks to a value indicating a posedge. This is
+  /// used to not create an excessive number of posedge detectors.
+  DenseMap<Value, Value> loweredPosedges;
+  /// The previous enable and the value it was lowered to. This is used to reuse
+  /// previous if ops for the same enable value.
+  std::pair<Value, Value> prevEnable;
+  /// The previous reset and the value it was lowered to. This is used to reuse
+  /// previous uf ops for the same reset value.
+  std::pair<Value, Value> prevReset;
```

Comment:
If there is more than one reset in a module and we process operations in an order where op1 uses resetA, op2 uses resetB, op3 uses resetA, etc., it would clone the reset fan-in each time?

Reply:
This just opportunistically remembers the last enable and reset that was lowered, in order to get a chance to fuse `scf.if` ops on the fly. You could delete all this and would get correct output as well, just with more if ops. The "lowering" here is also pretty trivial: it's mostly just `arc.state_read` operations that are reused. There is no more copying of fan-ins happening.